src/charm.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charmed Kubernetes Operator for the PostgreSQL database."""
+
+import logging
+import secrets
+import string
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus
+from ops.pebble import Layer
+
+logger = logging.getLogger(__name__)
+
+
+class PostgresqlOperatorCharm(CharmBase):
+    """Charmed Operator for the PostgreSQL database."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self._postgresql_service = "postgresql"
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.postgresql_pebble_ready, self._on_postgresql_pebble_ready)
+
+    def _on_install(self, _):
+        """Event handler for InstallEvent."""
+        # TODO: change to peer/leader data bag when relations are implemented.
+        self._stored.postgres_password = self._new_password()
+
+    def _on_config_changed(self, _):
+        """Handle the config-changed event."""
+        # TODO: placeholder method to implement logic specific to configuration change.
+        pass
+
+    def _on_postgresql_pebble_ready(self, _) -> None:
+        """Event handler for on PebbleReadyEvent."""
+        # TODO: move this code to an "_update_layer" method in order to also utilize it in
+        # config-changed hook.
+        # Get the postgresql container so we can configure/manipulate it.
+        container = self.unit.get_container(self._postgresql_service)
```

Comment:
I'm unsure about this one. If you deploy the application with a different name, this might not work. Something more generic could be a better option:

```python
self.unit.get_container(self.model.app.name)
```

It can also help differentiate between the name of the service and the container.

Reply:
Good suggestion. Thanks, Daniela. Testing, seems like that using the app name also may cause the charm to not find the container if we use a different application name.

Maybe the best approach should be using the `workload` property from the `event` argument. What do you think @jnsgruk? (I saw `event.workload` in the Kubernetes dashboard charm) Maybe it can be something like:

```python
def _on_postgresql_pebble_ready(self, event: WorkloadEvent) -> None:
    """Event handler for on PebbleReadyEvent."""
    container = event.workload
```

Reply:
Yep, you can definitely do that inside the `pebble-ready` events -- obviously doesn't work in any other kind of event handler but an easy way to make sure you're getting the right container here :)

Reply:
Great. I updated the code to use `event.workload`. Thanks for the note about the use only on `pebble-ready` event.